### ISSUE TITLE
Flatpak uses icons from host instead of the flatpak-packaged ones

### ISSUE THREAD

**Issue by jaje** [flatpak/flatpak#2149]
## Linux distribution and version 
Fedora 29
## Flatpak version
Flatpak 1.0.2
## Description of the problem
Flatpak uses icons from host instead of the ones from org.freedesktop.Platform.Icontheme.*.
This causes crash in gnome builder and vscode.
## Steps to reproduce
- Inside fedora install Paper theme
- Install org.freedesktop.Platform.Icontheme.Paper
- `flatpak run org.gnome.Builder`:
```
(gnome-builder:2): Gtk-WARNING **: 20:52:21.780: Could not load a pixbuf from icon theme.
This may indicate that pixbuf loaders or the mime database could not be found.
**
Gtk:ERROR:gtkiconhelper.c:494:ensure_surface_for_gicon: assertion failed (error == NULL): Failed to load /run/host/share/icons/Paper/16x16/status/image-missing.png: Error opening file /run/host/share/icons/Paper/16x16/status/image-missing.png: No such file or directory (g-io-error-quark, 1)
```
After that gnome builder crashes
- vscode crashes when opening folder with similar error
 
Output from `file` from host system:
`/usr/share/icons/Paper/16x16/status/image-missing.png: symbolic link to /usr/share/icons/Paper/16x16/status/gtk-missing-image.png`

When I reorder folders in `XDG_DATA_DIRS` (like `flatpak run --env=XDG_DATA_DIRS=/run/host/share:/app/share:/usr/share:/usr/share/runtime/share org.gnome.Builder`) then it works without problem.

**Comment by matthiasclasen** [comment]
I don't follow. the reordered XDG_DATA_DIRS look like they will make the app use the host icons instead of the extension ones. What gives ?

**Comment by jaje** [comment]
I have just tried other XDG_DATA_DIRS:
- `flatpak run --env=XDG_DATA_DIRS=/app/share:/usr/share:/usr/share/runtime/share org.gnome.Builder`: works without problem (host path removed from XDG_DATA_DIRS)

- `flatpak run --env=XDG_DATA_DIRS=/app/share:/run/host/share:/usr/share:/usr/share/runtime/share org.gnome.Builder`: works, but with warning:
```
(gnome-builder:2): Gtk-WARNING **: 10:26:56.251: Could not load a pixbuf from icon theme.
This may indicate that pixbuf loaders or the mime database could not be found.
```
- `flatpak run --env=XDG_DATA_DIRS=/app/share:/usr/share:/run/host/share:/usr/share/runtime/share org.gnome.Builder`: the same as above

Output from `file /run/host/share/icons/Paper/16x16/status/image-missing.png` from inside flatpak:
```
/run/host/share/icons/Paper/16x16/status/image-missing.png: broken symbolic link to /usr/share/icons/Paper/16x16/status/gtk-missing-image.png
```

**Comment by matthiasclasen** [comment]
This is a bug in the Paper icontheme. That absolute symlink makes it not relocatable.

**Comment by jaje** [comment]
Yes, it is, but when Paper extension is installed it should not matter if Paper theme from host is broken.

**Comment by matthiasclasen** [comment]
Well, the truth of the matter is: after we had already defined the icontheme extensions, we figured that it is generally safe to just use host icons. You found a case where this breaks down: if the host icontheme is not relocatable

**Comment by alexlarsson** [comment]
The question is if we should allow the host theme to override the runtime theme (be it via an extension or not). We believed this to be safe, but clearly there are cases where its not, as this shows. 

However, I still think its the right thing to do. Themes are essentially a user controlled thing, and the user should be in control of it, even if doing so is risky (themes were always risky in some ways). And full control implies to me following the host theme. Making the paper extension override the host version would fix it only for a subset of users that had it installed anyway, so its not a full fix.

**Comment by alexlarsson** [comment]
Still, its bad if apps crash. We could possibly scan the icon theme dir before using it and warn if we find absolute symlinks and disable that dir?

**Comment by e-minguez** [comment]
BZ opened to fix paper-icon-theme rpm https://bugzilla.redhat.com/show_bug.cgi?id=1639717

**Comment by decathorpe** [comment]
I'm maintaining the `paper-icon-theme` package for fedora, and I'm only now finding out about this issue. I've never had flatpak applications crash on me because of an icon theme (I'm using Paper myself, too), but it's clearly happening in this case.

It looks like this problem arises from several things going slightly wrong.

1. The Paper icon theme upstream has defined symlinks for duplicate icons, but
2. `meson` will copy the underlying file instead of copying the symlink (for now, see below), and
3. fedora packaging tries to de-duplicate the files duplicated by meson, because otherwise the package is twice the size (about 90 MB instead of 45 MB).

I've tried switching the de-duplication process to use hardlinks instead of symlinks, which seemed to work (on rawhide), but it crashed `fdupes` with a SIGSEGV when I tried to build the package on fedora 28. So I'm not really sure how to proceed.

And, to make things interesting, the whole thing will blow up again once meson changes its behavior, as this is the warning that's printed when copying symlinks:

```
Warning: trying to copy a symlink that points to a file. This will copy the file,
but this will be changed in a future version of Meson to copy the symlink as is. Please update your
build definitions so that it will not break when the change happens.
```

**Comment by matthiasclasen** [comment]
So the concrete suggestion here is to check that host directories are relocatable (don't contain absolute symlinks, anything else?) before mounting them in the sandbox

**Comment by smcv** [comment]
> Could the symlinks just be relative instead?

For what it's worth, Debian policy says symlinks within a top-level directory (like `/usr`) should be relative, and symlinks between top-level directories (for example between `/etc` and `/var`) should be absolute. The `dh_link` helper tool fixes symlinks in the temporary `DESTDIR` during package build, to make them conform to this policy. I'm sure other distributions would be welcome to copy that bit of policy and/or its implementation.

**Comment by decathorpe** [comment]
I mean, I could just drop the de-duplication step from the Paper icon theme package in fedora. It would blow up the installed package size by about 40%, but there wouldn't be any symlinks anymore.

**Comment by TingPing** [comment]
@decathorpe 	Just make them relative symlinks and it probably works.

**Comment by decathorpe** [comment]
@TingPing As I mentioned above, the script that's used for deduplication doesn't support that.
It was added by the previous maintanier of the package, and it reads like dark magic to me.

I'll probably just drop the deduplication altogether, since it causes issues like this one.

**Comment by phillipsj** [comment]
Just wanted to post I am seeing the same exact issue on Fedora 29.

flatpak run io.github.Pithos

(pithos:2): Gtk-WARNING **: 20:07:41.529: Could not load a pixbuf from icon theme.
This may indicate that pixbuf loaders or the mime database could not be found.
**
Gtk:ERROR:gtkiconhelper.c:494:ensure_surface_for_gicon: assertion failed (error == NULL): Failed to load /run/host/share/icons/Paper/16x16/status/image-missing.png: Error opening file /run/host/share/icons/Paper/16x16/status/image-missing.png: No such file or directory (g-io-error-quark, 1)

**Comment by decathorpe** [comment]
@phillipsj Can you check if this update https://bodhi.fedoraproject.org/updates/FEDORA-2019-6e40282ec8 fixes the issue for you? It's available from the updates-testing repository now, and should be available from the stable fedora repository tomorrow.

**Comment by jp7677** [comment]
@decathorpe All good here again. thanks a lot!

**Comment by decathorpe** [comment]
So, we worked around this specific instance of this issue, but the generic problem remains - am I correct?
